comb_spec_searcher/strategies/constructor.py

```diff
@@ -422,6 +430,235 @@ def __str__(self) -> str:
+        self._min_sizes = tuple(child.minimum_size_of_object() for child in children)
+        self._max_sizes = tuple(
+            child.minimum_size_of_object() if child.is_atom() else None
+            for child in children
+        )
+        self._parent_param_map = self._build_parent_param_map(parent, children)
+        self._parent_shift = sum(self._min_sizes) - self._min_sizes[self.idx]
+        self._num_parent_params = len(parent.extra_parameters)
+
+    @staticmethod
+    def build_param_map(
+        child_pos_to_parent_pos: Tuple[Tuple[int, ...], ...], num_parent_params: int
+    ) -> ParametersMap:
+        def param_map(param: Parameters) -> Parameters:
+            new_params: List[Optional[int]] = [None for _ in range(num_parent_params)]
+            for pos, value in enumerate(param):
+                parent_pos = child_pos_to_parent_pos[pos]
+                for p in parent_pos:
+                    assert new_params[p] is None or new_params[p] == value
+                    new_params[p] = value
+            assert all(p is not None for p in new_params)
+            return cast(Parameters, tuple(new_params))
+
+        return param_map
+
+    def _build_parent_param_map(
+        self,
+        parent: CombinatorialClassType,
+        children: Tuple[CombinatorialClassType, ...],
+    ) -> ParametersMap:
+        extra_param = self.extra_parameters[self.idx]
+        child = children[self.idx]
+        child_param_to_pos = {
+            param: pos for pos, param in enumerate(child.extra_parameters)
+        }
+        parent_pos_to_child_pos: Tuple[Tuple[int, ...], ...] = tuple(
+            (child_param_to_pos[extra_param[parent_var]],)
+            if parent_var in extra_param
+            else tuple()
+            for parent_var in parent.extra_parameters
+        )
+        return self.build_param_map(
+            parent_pos_to_child_pos, len(child.extra_parameters)
+        )
+
+    def _new_param(self, *children_params: Parameters) -> Parameters:
+        """
+        Computes the parameter values on the parent that the given children parameters
+        contribute to.
+        """
+        mapped_params = (
+            pmap(p) for pmap, p in zip(self._children_param_maps, children_params)
+        )
+        return tuple(sum(vals) for vals in zip(*mapped_params))
+
+    def _other_new_param(self, *children_params: Parameters) -> Parameters:
+        """
+        Computes the parameter values on the parent that the given children parameters
+        contribute to.
+        """
+        mapped_params = (
+            pmap(p)
+            for pmap, p in zip(
+                (m for i, m in enumerate(self._children_param_maps) if i != self.idx),
+                children_params,
+            )
+        )
+        return tuple(sum(vals) for vals in zip(*mapped_params))
+
+    def _a(
+        self,
+        n: int,
+        parent_subterm: Callable[[int], Terms],
+        children_subterms: SubTerms,
+    ) -> Terms:
```

Comment:
Need to add some docstring about what are the a,b and c dict. I'll add a small explaination of how it works in the class docstring I think.